Consolidate pragma section

BillWagner opened the pull request:
Fixes #23320 

Fixes other issues in this area while doing these edits:

Fixes #16763
Fixes #20414

This may be easiest to review commit by commit:

1. Fix other issues in the section
1. Combine all preprocessor files.
1. First pass at formatting.
1. Edit the single doc for content.
1. Grammar review on single doc.
1. Update all links.

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives?branch=pr-en-us-23374)

BillWagner committed:
add text for existing issues

There are two existing issues in this section of content. This commit fixes both.

BillWagner committed:
combine all files.

BillWagner committed:
first pass at formatting.

Youssef1313 commented on docs/csharp/language-reference/preprocessor-directives.md:
```diff
@@ -0,0 +1,428 @@
+description: "Learn the different C# preprocessor directives that control conditional compilation, warnings, nullable analysis, and more"
+title: "C# preprocessor directives"
+ms.date: 03/17/2021
+f1_keywords: 
+  - "cs.preprocessor"
```
f1 keywords from the deleted articles are now dropped.

BillWagner replied:
Thanks @Youssef1313 That's a great catch. I'll add them to the new page today.

BillWagner committed:
restore F1 and helpviewer keywords.

BillWagner committed:
fix build warnings.

BillWagner committed:
organize by scenario, not keyword.

BillWagner committed:
first pass edit.

BillWagner committed:
finish proofreading.

BillWagner committed:
fix warning

BillWagner committed:
run clean-repo to replace links

First run: replace redirected file links.

BillWagner committed:
Replace relative links

Run clean repo to replace absolute links (introduced in the previous run) with relative links.

BillWagner committed:
Revert "Replace relative links"

This reverts commit d3c931e2bf9f3fc17832ef176a7dc3818fff70b5.

BillWagner committed:
run with the new cleanrepo build

This way, forward slashes aren't replaced with backslashes.

BillWagner commented:
Closing and reopening for a fresh build.

tdykstra commented on docs/csharp/language-reference/preprocessor-directives.md:
```diff
@@ -0,0 +1,351 @@
+
+`#if`, along with the `#else`, `#elif`, `#endif`, `#define`, and `#undef` directives, lets you include or exclude code based on the existence of one or more symbols. Conditional compilation can be useful when compiling code for a debug build or when compiling for a specific configuration.
+
+A conditional directive beginning with a `#if` directive must explicitly be terminated with a `#endif` directive. `#define` lets you define a symbol. By using the symbol as the expression passed to the `#if` directive, the expression evaluates to `true`. You can also define a symbol with the [**DefineConstants**](compiler-options/language.md#defineconstants) compiler option. You can undefine a sy ...
```
```suggestion
A conditional directive beginning with an `#if` directive must explicitly be terminated with an `#endif` directive. `#define` lets you define a symbol. By using the symbol as the expression passed to the `#if` directive, the expression evaluates to `true`. You can also define a symbol with the [**DefineConstants**](compiler-options/language.md#defineconstants) compiler option. You can undefine a symbol with `#undef`. The scope of a symbol created with `#define` is the file in which it was defined. A symbol that you define with **DefineConstants** or with `#define` doesn't conflict with a variable of the same name. That is, a variable name shouldn't be passed to a preprocessor directive, and a symbol can only be evaluated by a preprocessor directive.
```

tdykstra commented on docs/csharp/language-reference/preprocessor-directives.md:
```diff
@@ -0,0 +1,351 @@
+
+A conditional directive beginning with a `#if` directive must explicitly be terminated with a `#endif` directive. `#define` lets you define a symbol. By using the symbol as the expression passed to the `#if` directive, the expression evaluates to `true`. You can also define a symbol with the [**DefineConstants**](compiler-options/language.md#defineconstants) compiler option. You can undefine a sy ...
+
+`#elif` lets you create a compound conditional directive. The `#elif` expression will be evaluated if neither the preceding `#if` nor any preceding, optional, `#elif` directive expressions evaluate to `true`. If a `#elif` expression evaluates to `true`, the compiler evaluates all the code between the `#elif` and the next conditional directive. For example:
```
```suggestion
`#elif` lets you create a compound conditional directive. The `#elif` expression will be evaluated if neither the preceding `#if` nor any preceding, optional, `#elif` directive expressions evaluate to `true`. If an `#elif` expression evaluates to `true`, the compiler evaluates all the code between the `#elif` and the next conditional directive. For example:
```

tdykstra commented on docs/csharp/language-reference/preprocessor-directives.md:
````diff
@@ -0,0 +1,351 @@
+#pragma checksum "filename" "{guid}" "checksum bytes"
+```
+
+Where `"filename"` is the name of the file that requires monitoring for changes or updates, `"{guid}"` is the Globally Unique Identifier (GUID) for the hash algorithm, and `"checksum_bytes"` is the string of hexadecimal digits representing the bytes of the checksum. Must be an even number of hexadecimal digits. An odd number of digits results in a compile-time warning, and the directive are ignor ...
````
```suggestion
Where `"filename"` is the name of the file that requires monitoring for changes or updates, `"{guid}"` is the Globally Unique Identifier (GUID) for the hash algorithm, and `"checksum_bytes"` is the string of hexadecimal digits representing the bytes of the checksum. Must be an even number of hexadecimal digits. An odd number of digits results in a compile-time warning, and the directive is ignored.
```

tdykstra reviewed (approved):
Looks great, just nit suggestions.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Tom Dykstra <tdykstra@microsoft.com>

BillWagner committed:
respond to code review comments.

Youssef1313 commented on docs/architecture/porting-existing-aspnet-apps/strategies-migrating-in-production.md:
```diff
@@ -37,7 +37,7 @@ Once the facade is in place, you can route part of it to a new ASP.NET Core app.
 ## Multi-targeting approaches
 
-Large apps that target .NET Framework may be migrated to ASP.NET Core over time by using multi-targeting and separate code paths for each framework. For example, code that must run in both environments could be modified with [preprocessor `#if`](../../csharp/language-reference/preprocessor-directives/preprocessor-if.md) directives to implement different functionality or use different dependencies ...
+Large apps that target .NET Framework may be migrated to ASP.NET Core over time by using multi-targeting and separate code paths for each framework. For example, code that must run in both environments could be modified with [preprocessor `#if`](../../csharp/language-reference/preprocessor-directives.md) directives to implement different functionality or use different dependencies when run in .NE ...
```
@BillWagner It would be a good addition to have these links point to the exact heading.

BillWagner replied:
Good point @Youssef1313 

I'll open a fresh PR for that.
